subsys/bluetooth/host/audio/iso.c

```diff
@@ -0,0 +1,1075 @@
+						 size_t reserve,
+						 k_timeout_t timeout,
+						 const char *func, int line)
+#else
+struct net_buf *bt_iso_create_frag_timeout(size_t reserve, k_timeout_t timeout)
+#endif
+{
+	struct net_buf_pool *pool = NULL;
+
+#if CONFIG_BT_L2CAP_TX_FRAG_COUNT > 0
+	pool = &iso_frag_pool;
+#endif
+
+#if defined(CONFIG_NET_BUF_LOG)
+	return bt_conn_create_pdu_timeout_debug(pool, reserve, timeout, func,
+						line);
+#else
+	return bt_conn_create_pdu_timeout(pool, reserve, timeout);
+#endif
+}
+
+static struct net_buf *hci_le_set_cig_params(struct bt_iso_create_param *param)
+{
+	struct bt_hci_cis_params *cis;
+	struct bt_hci_cp_le_set_cig_params *req;
+	struct net_buf *buf;
+	struct net_buf *rsp;
+	int i, err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_CIG_PARAMS,
+				sizeof(*req) + sizeof(*cis) * param->num_conns);
+	if (!buf) {
+		return NULL;
+	}
+
+	req = net_buf_add(buf, sizeof(*req));
+
+	memset(req, 0, sizeof(*req));
+
+	req->cig_id = param->conns[0]->iso.cig_id;
+	sys_put_le24(param->chans[0]->qos->interval, req->m_interval);
+	sys_put_le24(param->chans[0]->qos->interval, req->s_interval);
+	req->sca = param->chans[0]->qos->sca;
+	req->packing = param->chans[0]->qos->packing;
+	req->framing = param->chans[0]->qos->framing;
+	req->m_latency = sys_cpu_to_le16(param->chans[0]->qos->latency);
+	req->s_latency = sys_cpu_to_le16(param->chans[0]->qos->latency);
+	req->num_cis = param->num_conns;
+
+	/* Program the cis parameters */
+	for (i = 0; i < param->num_conns; i++) {
+		cis = net_buf_add(buf, sizeof(*cis));
+
+		memset(cis, 0, sizeof(*cis));
+
+		cis->cis_id = param->conns[i]->iso.cis_id;
+
+		switch (param->chans[i]->qos->dir) {
+		case BT_ISO_CHAN_QOS_IN:
+			cis->m_sdu = param->chans[i]->qos->sdu;
+			break;
+		case BT_ISO_CHAN_QOS_OUT:
+			cis->s_sdu = param->chans[i]->qos->sdu;
+			break;
+		case BT_ISO_CHAN_QOS_INOUT:
+			cis->m_sdu = param->chans[i]->qos->sdu;
+			cis->s_sdu = param->chans[i]->qos->sdu;
+			break;
+		}
+
+		cis->m_phy = param->chans[i]->qos->phy;
+		cis->s_phy = param->chans[i]->qos->phy;
+		cis->m_rtn = param->chans[i]->qos->rtn;
+		cis->s_rtn = param->chans[i]->qos->rtn;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_CIG_PARAMS, buf, &rsp);
+	if (err) {
+		return NULL;
+	}
+
+	return rsp;
+}
+
+int bt_conn_bind_iso(struct bt_iso_create_param *param)
+{
+	struct bt_conn *conn;
+	struct net_buf *rsp;
+	struct bt_hci_rp_le_set_cig_params *cig_rsp;
+	int i, err;
+
+	/* Check if controller is ISO capable */
+	if (!BT_FEAT_LE_CIS_MASTER(bt_dev.le.features)) {
+		return -ENOTSUP;
+	}
+
+	if (!param->num_conns || param->num_conns > CONFIG_BT_MAX_ISO_CONN) {
+		return -EINVAL;
+	}
+
+	/* Assign ISO connections to each LE connection */
+	for (i = 0; i < param->num_conns; i++) {
+		conn = param->conns[i];
+
+		if (conn->type != BT_CONN_TYPE_LE) {
+			return -EINVAL;
+		}
+
+		conn = bt_conn_add_iso(conn);
+		if (!conn) {
+			return -ENOMEM;
+		}
+
+		/* TODO: Figure out if ID was already assigned? */
```

Comment:
Can't this be done with a simple for loop over existing ISOs?

Reply:
What this is doing is allocating a ISO (bt_conn) object for each ACL since that is then used by hci_le_set_cig_params, so there is no existing ISOs yet, or the user want to create new connections anyway.

Reply:
If there are no existing ISOs yet, do we need to check if IDs was assigned here? In other words, can the TODO be removed?

Reply:
The TODO is for the CIG id which may have been set by the application so I think we will need to lookup if there is already in use. I do intend to add auto allocate logic though, in case the application don't care to use a specific CIG id.